notebook/gateway/managers.py

```diff
@@ -530,38 +605,86 @@ def _get_kernelspecs_endpoint_url(self, kernel_name=None):
+        """
+
+        kernel_mlnode = {}
+        _kernel_specs = []
+
+
+        # Fetching all the kernels from the available MLnodes.
+
+        for endpoint in self.base_endpoints:
+            response = yield gateway_request('http://' + endpoint, method='GET')
+            kernel_mlnode[endpoint] = json_decode(response.body)
+
+        """
+        Since the data structure that is send by mlnodes is same. When sending the data to FE the data structure
+        would be updated. Avoiding that by creating a unique keys. Assuming the name of mlnode to be unique.
+        For example:
+        {'168.62.201.192:8888/api/kernelspecs': 
+            {'default': 'python3', 'kernelspecs': 
+                {'python3': 
+                    {'name': 'python3', 'spec': 
+                                        {'argv': 
+                                            ['/opt/conda/bin/python', '-m', 'ipykernel_launcher', '-f', 
+                                            '{connection_file}'], 'env': {}, 'display_name': 'Python 3', 
+                                            'language': 'python', 'interrupt_mode': 'signal', 'metadata': {}}, 
+                                            'resources': {'logo-64x64': '/kernelspecs/python3/logo-64x64.png', 
+                                            'logo-32x32': '/kernelspecs/python3/logo-32x32.png'}}, 'spylon-kernel': 
+                                            {'name': 'spylon-kernel', 'spec': {'argv': ['/opt/conda/bin/python', '-m', 
+                                            'spylon_kernel', '-f', '{connection_file}'], 
+                                            'env': {'PYTHONUNBUFFERED': '1', 'SPARK_SUBMIT_OPTS': 
+                                            '-Dscala.usejavacp=true'}, 'display_name': 'spylon-kernel', 
+                                            'language': 'scala', 'interrupt_mode': 'signal', 'metadata': {}},
+                                             'resources': {'logo-64x64': '/kernelspecs/spylon-kernel/logo-64x64.png', 
+                                             'logo-32x32': '/kernelspecs/spylon-kernel/logo-32x32.png'}}, 
+                                             'ir': {'name': 'ir', 'spec': {'argv': ['R', '--slave', '-e', 
+                                             'IRkernel::main()', '--args', '{connection_file}'], 'env': {}, 
+                                             'display_name': 'R', 'language': 'R', 'interrupt_mode': 'signal', 
+                                             'metadata': {}}, 'resources': {'kernel.js': '/kernelspecs/ir/kernel.js', 
+                                             'logo-64x64': '/kernelspecs/ir/logo-64x64.png'}
+                                        }
+                    }
+            }
+        }
+        """
+
+        for key_mlnode, value_mlnode in kernel_mlnode.items():
+            __kernel_specs = {'kernelspecs': {}}
+            __kernel_specs__ = value_mlnode['kernelspecs']
```

Comment:
curious, in terms of naming patterns, two dunder signs imply private variables to the class. Why is that naming convention required here?

Reply:
because they are private variables not being used outside the scope of the function.